collection.go

```diff
@@ -366,23 +362,105 @@ func (c *collection) Get(key Reference) (Document, error) {
 }
 
 func (c *collection) documentBucket(tx *bbolt.Tx) *bbolt.Bucket {
-	bucket := tx.Bucket([]byte(c.Name))
+	bucket := tx.Bucket([]byte(c.name))
 	if bucket == nil {
 		return nil
 	}
 	return bucket.Bucket(documentCollectionByteRef())
 }
 
-// ValuesAtPath returns a slice with the values found at the given JSON path query
-func (c *collection) ValuesAtPath(document Document, jsonPath string) ([]Scalar, error) {
+// valueCollector is responsible for going through the document and finding the Scalars that match the Query
+type valueCollector func(collection *collection, document Document, queryPath QueryPath) ([]Scalar, error)
+
+// JSONPathValueCollector collects values at a given JSON path expression. Objects are delimited by a dot and lists use an extra # in the expression:
+// object.list.#.key
+func JSONPathValueCollector(_ *collection, document Document, queryPath QueryPath) ([]Scalar, error) {
+	jsonPath, ok := queryPath.(jsonPath)
+	if !ok {
+		return nil, ErrInvalidQuery
+	}
+
 	if !gjson.ValidBytes(document) {
 		return nil, ErrInvalidJSON
 	}
-	result := gjson.GetBytes(document, jsonPath)
+	result := gjson.GetBytes(document, string(jsonPath))
 
 	return valuesFromResult(result)
 }
 
+// JSONLDValueCollector collects values given a list of IRIs that represent the nesting of the objects.
+func JSONLDValueCollector(collection *collection, document Document, queryPath QueryPath) ([]Scalar, error) {
+	termPath, ok := queryPath.(termPath)
+	if !ok {
+		return nil, ErrInvalidQuery
+	}
+
+	if len(termPath.terms) == 0 {
+		return []Scalar{}, nil
+	}
+
+	var input interface{}
+	if err := json.Unmarshal(document, &input); err != nil {
+		return nil, err
+	}
+
+	options := ld.NewJsonLdOptions("")
+	options.DocumentLoader = collection.documentLoader
+	expanded, err := ld.NewJsonLdProcessor().Expand(input, options)
+	if err != nil {
+		return nil, err
+	}
+
+	return valuesFromSliceAtPath(expanded, termPath), nil
+}
+
+func valuesFromSliceAtPath(expanded []interface{}, termPath termPath) []Scalar {
+	result := make([]Scalar, 0)
+
+	for _, sub := range expanded {
+		switch typedSub := sub.(type) {
+		case []interface{}:
+			result = append(result, valuesFromSliceAtPath(typedSub, termPath)...)
+		case map[string]interface{}:
+			result = append(result, valuesFromMapAtPath(typedSub, termPath)...)
+		}
+	}
+
+	return result
+}
+
+func valuesFromMapAtPath(expanded map[string]interface{}, termPath termPath) []Scalar {
+	// JSON-LD in expanded form either has @value, @id, @list or @set
+	if termPath.IsEmpty() {
+		if value, ok := expanded["@value"]; ok {
+			return []Scalar{MustParseScalar(value)}
+		}
+		if id, ok := expanded["@id"]; ok {
+			return []Scalar{MustParseScalar(id)}
+		}
+		if list, ok := expanded["@list"]; ok {
+			castList := list.([]interface{})
+			return valuesFromSliceAtPath(castList, termPath)
+		}
```

Comment:
what about `@set` then?

Reply:
`@set` translates into a list of `@value`s